wolfcrypt/src/rsa.c

```diff
@@ -1134,42 +1138,47 @@ static int RsaPublicEncryptEx(const byte* in, word32 inLen, byte* out,
     switch (key->state) {
     case RSA_STATE_NONE:
     case RSA_STATE_ENCRYPT_PAD:
-
-    #if defined(WOLFSSL_ASYNC_CRYPT) && defined(HAVE_CAVIUM)
-        if (key->asyncDev.marker == WOLFSSL_ASYNC_MARKER_RSA) {
-            if (rsa_type == RSA_PUBLIC_ENCRYPT && pad_value == RSA_BLOCK_TYPE_2) {
-                key->state = RSA_STATE_ENCRYPT_RES;
-                key->tmpLen = key->n.used;
-                return NitroxRsaPublicEncrypt(in, inLen, out, outLen, key);
+        key->state = RSA_STATE_ENCRYPT_PAD;
+
+    #if defined(WOLFSSL_ASYNC_CRYPT) && defined(WC_ASYNC_ENABLE_RSA) && \
+            defined(HAVE_CAVIUM)
+        if (key->asyncDev.marker == WOLFSSL_ASYNC_MARKER_RSA && key->n.raw.buf) {
+            /* Async operations that include padding */
+            if (rsa_type == RSA_PUBLIC_ENCRYPT &&
+                                                pad_value == RSA_BLOCK_TYPE_2) {
+                key->state = RSA_STATE_ENCRYPT_EXPTMOD;
+                key->dataLen = key->n.raw.len;
+                ret = NitroxRsaPublicEncrypt(in, inLen, out, outLen, key);
+                break;
             }
-            else if (rsa_type == RSA_PRIVATE_ENCRYPT && pad_value == RSA_BLOCK_TYPE_1) {
-                key->state = RSA_STATE_ENCRYPT_RES;
-                key->tmpLen = key->n.used;
-                return NitroxRsaSSL_Sign(in, inLen, out, outLen, key);
+            else if (rsa_type == RSA_PRIVATE_ENCRYPT &&
+                                                pad_value == RSA_BLOCK_TYPE_1) {
+                key->state = RSA_STATE_ENCRYPT_EXPTMOD;
+                key->dataLen = key->n.raw.len;
+                ret = NitroxRsaSSL_Sign(in, inLen, out, outLen, key);
+                break;
             }
         }
     #endif
 
-        key->state = RSA_STATE_ENCRYPT_EXPTMOD;
-
-        ret = wc_RsaPad_ex(in, inLen, out, sz, pad_value, rng,
-                               pad_type, hash, mgf, label, labelSz, key->heap);
+        ret = wc_RsaPad_ex(in, inLen, out, sz, pad_value, rng, pad_type, hash,
+                                                mgf, label, labelSz, key->heap);
         if (ret < 0) {
             break;
         }
         /* fall through */
     case RSA_STATE_ENCRYPT_EXPTMOD:
-        key->state = RSA_STATE_ENCRYPT_RES;
+        key->state = RSA_STATE_ENCRYPT_EXPTMOD;
```

Comment:
Putting the assignment above the case statement makes more sense as key->state should be RSA_STATE_ENCRYPT_EXPTMOD before executing the code below the case statement.
I've seen this (setting after case) in many places.